File: ref/bitunix-demo/Demo/Golang/open_api_ws_sign.go
package main

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"math/rand"
	"time"
)

// generateNonce generates a random string as nonce
func generateNonce() string {
	const charset = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
	const length = 32
	b := make([]byte, length)
	for i := range b {
		b[i] = charset[rand.Intn(len(charset))]
	}
	return string(b)
}

// generateTimestamp generates the current timestamp
func generateTimestamp() string {
	return fmt.Sprintf("%d", time.Now().Unix())
}

// sha256Hex calculates the SHA256 hash of the input string
func sha256Hex(input string) string {
	hash := sha256.Sum256([]byte(input))
	return hex.EncodeToString(hash[:])
}

// generateSign generates authentication signature
func generateSign(nonce, timestamp, apiKey, secretKey string) string {
	digestInput := nonce + timestamp + apiKey
	digest := sha256Hex(digestInput)
	signInput := digest + secretKey
	return sha256Hex(signInput)
}

// WSAuth defines the WebSocket authentication data structure
type WSAuth struct {
	APIKey    string `json:"apiKey"`
	Timestamp int64  `json:"timestamp"`
	Nonce     string `json:"nonce"`
	Sign      string `json:"sign"`
}

// GetAuthWSFuture generates WebSocket authentication data
func GetAuthWSFuture(apiKey, secretKey string) WSAuth {
	nonce := generateNonce()
	timestamp := generateTimestamp()
	sign := generateSign(nonce, timestamp, apiKey, secretKey)

	// Convert timestamp string to int64
	timestampInt, _ := time.Parse(time.RFC3339, timestamp)

	return WSAuth{
		APIKey:    apiKey,
		Timestamp: timestampInt.Unix(),
		Nonce:     nonce,
		Sign:      sign,
	}
}

File: ref/bitunix-demo/Demo/PHP/open_api_ws_sign.php
<?php

/**
 * generate a random string as nonce
 * @return string
 */
function generate_nonce() {
    $characters = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789';
    $nonce = '';
    for ($i = 0; $i < 32; $i++) {
        $nonce .= $characters[rand(0, strlen($characters) - 1)];
    }
    return $nonce;
}

/**
 * generate current timestamp
 * @return string
 */
function generate_timestamp() {
    return (string)time();
}

/**
* calculate the SHA256 hash value of the input string
 * @param string $input_string
 * @return string
 */
function sha256_hex($input_string) {
    return hash('sha256', $input_string);
}

/**
 * generate authentication signature
 * @param string $nonce
 * @param string $timestamp
 * @param string $api_key
 * @param string $secret_key
 * @return string
 */
function generate_sign($nonce, $timestamp, $api_key, $secret_key) {
    $digest_input = $nonce . $timestamp . $api_key;
    $digest = sha256_hex($digest_input);
    $sign_input = $digest . $secret_key;
    return sha256_hex($sign_input);
}

/**
 * generate WebSocket authentication data
 * @param string $api_key
 * @param string $secret_key
 * @return array
 */
function get_auth_ws_future($api_key, $secret_key) {
    $nonce = generate_nonce();
    $timestamp = generate_timestamp();
    $sign = generate_sign($nonce, $timestamp, $api_key, $secret_key);
    
    return [
        'apiKey' => $api_key,
        'timestamp' => (int)$timestamp,
        'nonce' => $nonce,
        'sign' => $sign
    ];
}
